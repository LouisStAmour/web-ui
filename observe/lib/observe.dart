// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

/**
 * *Warning*: this library is experimental, and APIs are subject to change.
 *
 * This library is used to observe changes to [Observable] types. It also
 * has helpers to implement [Observable] objects.
 *
 * For example:
 *
 *     class Monster extends ObservableMixin {
 *       int _health = 100;
 *       static const _HEALTH = 'health';
 *       get health => _health;
 *       set health(value) {
 *         _health = notifyChange(_HEALTH, _health, value);
 *       }
 *       void damage(int amount) {
 *         print('$this takes $amount damage!');
 *         health -= amount;
 *       }
 *       toString() => 'Monster';
 *
 *       // These methods are temporary until dart2js supports mirrors.
 *       getValue(key) {
 *         if (key == _HEALTH) return health;
 *         return null;
 *       }
 *       setValue(key, val) {
 *         if (key == _HEALTH) health = val;
 *       }
 *     }
 *
 *     main() {
 *       var obj = new Monster();
 *       obj.changes.listen((records) {
 *         print('Changes to $obj were: $records');
 *       });
 *       // Asynchronously schedules delivery of these changes
 *       obj.damage(10);
 *       obj.damage(20);
 *       print('done!');
 *     }
 */
library observe;

import 'dart:async';
import 'dart:collection';
import 'dart:math' as math;

// TODO(jmesserly): use libraries instead of parts.
part 'src/list_diff.dart';
part 'src/observe_path.dart';
part 'src/observable.dart';
part 'src/observable_box.dart';
part 'src/observable_list.dart';
part 'src/observable_map.dart';
