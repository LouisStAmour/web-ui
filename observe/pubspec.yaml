name: observe
author: "Web UI Authors <web-ui-dev@dartlang.org>"
homepage: https://github.com/dart-lang/web-ui
description: observable objects for data binding
version: 0.0.1
dependencies:
  unittest: ">=0.4.6"
environment:
  sdk: ">=0.4.6"
