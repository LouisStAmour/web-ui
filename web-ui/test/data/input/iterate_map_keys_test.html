<!DOCTYPE html>
<!--
Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
for details. All rights reserved. Use of this source code is governed by a
BSD-style license that can be found in the LICENSE file.
-->
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <script src="packages/web_ui/testing/testing.js"></script>
</head>
<body>
  Tests that we can iterate keys of a map and get change notifications.
  With a normal template:
  <div id="test1">{<template iterate="k in map.keys">{{k}}: {{map[k]}}, </template>}</div>
  With a template attribute:
  <div id="test2">{<div template iterate="k in map.keys">{{k}}: {{map[k]}}, </div>}</div>
  <script type="application/dart">
  import 'dart:async';
  import 'dart:html';
  import 'package:unittest/unittest.dart';
  import 'package:web_ui/observe.dart';

  @observable var map = toObservable({'a': 1, 'b': 2});

  main() {

    String getTestText(String id) {
      var text = document.query(id).text;
      // TODO(jmesserly): would be nice if it was easier to write this in
      // templates, e.g. $index or similar.
      text = text.replaceFirst(', }', '}');
      return text;
    }

    Timer.run(() {
      expect(getTestText('#test1'), map.toString());
      expect(getTestText('#test2'), map.toString());
      map['c'] = 3;

      Timer.run(() {
        expect(getTestText('#test1'), map.toString());
        expect(getTestText('#test2'), map.toString());
        map['d'] = 4;

        window.postMessage('done', '*');
      });
    });
  }
  </script>
</body>
</html>
