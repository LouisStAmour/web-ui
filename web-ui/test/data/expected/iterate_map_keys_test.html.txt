Content-Type: text/plain
layer at (0,0) size 800x600
  RenderView at (0,0) size 800x600
layer at (0,0) size 800x142
  RenderBlock {HTML} at (0,0) size 800x142
    RenderBody {BODY} at (8,8) size 784x126
      RenderBlock (anonymous) at (0,0) size 784x36
        RenderText {#text} at (0,0) size 760x36
          text run at (0,0) width 700: "Tests that we can iterate keys of a map and get change notifications. "
          text run at (700,0) width 60: "With a"
          text run at (0,18) width 160: "normal template:"
      RenderBlock {DIV} at (0,36) size 784x18
        RenderText {#text} at (0,0) size 10x18
          text run at (0,0) width 10: "{"
        RenderText {#text} at (10,0) size 10x18
          text run at (10,0) width 10: "a"
        RenderText {#text} at (20,0) size 20x18
          text run at (20,0) width 20: ": "
        RenderText {#text} at (40,0) size 10x18
          text run at (40,0) width 10: "1"
        RenderText {#text} at (50,0) size 20x18
          text run at (50,0) width 20: ", "
        RenderText {#text} at (70,0) size 10x18
          text run at (70,0) width 10: "b"
        RenderText {#text} at (80,0) size 20x18
          text run at (80,0) width 20: ": "
        RenderText {#text} at (100,0) size 10x18
          text run at (100,0) width 10: "2"
        RenderText {#text} at (110,0) size 20x18
          text run at (110,0) width 20: ", "
        RenderText {#text} at (130,0) size 10x18
          text run at (130,0) width 10: "c"
        RenderText {#text} at (140,0) size 20x18
          text run at (140,0) width 20: ": "
        RenderText {#text} at (160,0) size 10x18
          text run at (160,0) width 10: "3"
        RenderText {#text} at (170,0) size 20x18
          text run at (170,0) width 20: ", "
        RenderText {#text} at (190,0) size 10x18
          text run at (190,0) width 10: "d"
        RenderText {#text} at (200,0) size 20x18
          text run at (200,0) width 20: ": "
        RenderText {#text} at (220,0) size 10x18
          text run at (220,0) width 10: "4"
        RenderText {#text} at (230,0) size 20x18
          text run at (230,0) width 20: ", "
        RenderText {#text} at (250,0) size 10x18
          text run at (250,0) width 10: "}"
      RenderBlock (anonymous) at (0,54) size 784x18
        RenderText {#text} at (0,0) size 260x18
          text run at (0,0) width 260: "With a template attribute:"
      RenderBlock {DIV} at (0,72) size 784x54
        RenderBlock (anonymous) at (0,0) size 784x18
          RenderText {#text} at (0,0) size 10x18
            text run at (0,0) width 10: "{"
        RenderBlock {DIV} at (0,18) size 784x18
          RenderText {#text} at (0,0) size 10x18
            text run at (0,0) width 10: "a"
          RenderText {#text} at (10,0) size 20x18
            text run at (10,0) width 20: ": "
          RenderText {#text} at (30,0) size 10x18
            text run at (30,0) width 10: "1"
          RenderText {#text} at (40,0) size 20x18
            text run at (40,0) width 20: ", "
          RenderText {#text} at (60,0) size 10x18
            text run at (60,0) width 10: "b"
          RenderText {#text} at (70,0) size 20x18
            text run at (70,0) width 20: ": "
          RenderText {#text} at (90,0) size 10x18
            text run at (90,0) width 10: "2"
          RenderText {#text} at (100,0) size 20x18
            text run at (100,0) width 20: ", "
          RenderText {#text} at (120,0) size 10x18
            text run at (120,0) width 10: "c"
          RenderText {#text} at (130,0) size 20x18
            text run at (130,0) width 20: ": "
          RenderText {#text} at (150,0) size 10x18
            text run at (150,0) width 10: "3"
          RenderText {#text} at (160,0) size 20x18
            text run at (160,0) width 20: ", "
          RenderText {#text} at (180,0) size 10x18
            text run at (180,0) width 10: "d"
          RenderText {#text} at (190,0) size 20x18
            text run at (190,0) width 20: ": "
          RenderText {#text} at (210,0) size 10x18
            text run at (210,0) width 10: "4"
          RenderText {#text} at (220,0) size 10x18
            text run at (220,0) width 10: ","
        RenderBlock (anonymous) at (0,36) size 784x18
          RenderText {#text} at (0,0) size 10x18
            text run at (0,0) width 10: "}"
